### PR TITLE
Fix SWC dynamic transform with suspense but without ssr

### LINKED ISSUE #36636: dynamic import with `suspense:true, ssr:false` causes `The server could not finish this Suspense boundary` error
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

```
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 21.2.0: Sun Nov 28 20:28:41 PST 2021; root:xnu-8019.61.5~1/RELEASE_ARM64_T6000
Binaries:
  Node: 16.14.2
  npm: 8.1.4
  Yarn: 1.22.1
  pnpm: 6.32.4
Relevant packages:
  next: 12.1.6-canary.16
  react: 18.1.0
  react-dom: 18.1.0
```

### Describe the Bug

Using dynamic component import with `{suspense: true, ssr:false}` cause this error: 
```
Error: The server could not finish this Suspense boundary, likely due to an error during server rendering. Switched to client rendering.
```

image

**Important Note:** functionally this works correctly, the issue is just that an error is being generated and so it appears broken

### Expected Behavior

Should work without any error popup or error in browser console

#### Runable reproduction: https://codesandbox.io/s/dank-sound-pkxopr?file=/pages/index.js

```tsx
import { Suspense } from "react";
import dynamic from "next/dynamic";

const Thing = dynamic(() => import("../thing"), { ssr: false, suspense: true });

export default function IndexPage() {
  return (
    <div>
      <p>Next.js Example</p>
      <Suspense fallback="Loading...">
        <Thing />
      </Suspense>
    </div>
  );
}
```

```tsx
//thing.js
export default function Thing() {
  return "Thing";
}
```

### Other

You can also cause this same behavior with this next.js page:

```tsx
import { Suspense } from "react";

export default function Thing() {
  if (typeof window === 'undefined') {
    const e = new error()
    e.name = "rendering suspense fallback..."
    delete e.stack
    throw e
  } 
  return "Thing"
}

export default function IndexPage() {
  return (
    <div>
      <p>Next.js Example</p>
      <Suspense fallback="Loading...">
        <Thing />
      </Suspense>
    </div>
  );
}
```

### PR DESCRIPTION
**Pull request by Brooooooklyn**
The Babel plugin works fine, so it seems not a runtime issue.

fixes https://github.com/vercel/next.js/issues/36636

### CODE REVIEW

File: packages/next-swc/crates/core/src/next_dynamic.rs [hunk 1]
```diff
@@ -260,14 +261,25 @@ impl Fold for NextDynamicPatcher {
                                                 has_ssr_false = true
                                             }
                                         }
+                                        if sym == "suspense" {
+                                            if let Some(Lit::Bool(Bool {
+                                                value: true,
+                                                span: _,
+                                            })) = match &**value {
```
Comment (kdy1) [inline packages/next-swc/crates/core/src/next_dynamic.rs:268]: `value.as_lit()` seems better than a match expression in the rhs of if let I think

**Review by kdy1 — CHANGES_REQUESTED** [review]

File: packages/next-swc/crates/core/src/next_dynamic.rs (hunk 1 above)
Comment (kdy1) [reply packages/next-swc/crates/core/src/next_dynamic.rs:268]: This situation will be improved when https://github.com/rust-lang/rust/issues/87121 lands

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js dynamic-suspense-without-ssr | Change |
| - | - | - | - |
| buildDuration | 19.5s | 19.6s | ⚠️ +76ms |
| buildDurationCached | 7.6s | 7.6s | ⚠️ +34ms |
| nodeModulesSize | 475 MB | 475 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js dynamic-suspense-without-ssr | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 5.014 | 4.944 | -0.07 |
| / avg req/sec | 498.62 | 505.7 | +7.08 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.708 | 1.782 | ⚠️ +0.07 |
| /error-in-render avg req/sec | 1463.99 | 1402.88 | ⚠️ -61.11 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js dynamic-suspense-without-ssr | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 28.8 kB | 28.8 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 72.5 kB | 72.5 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js dynamic-suspense-without-ssr | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js dynamic-suspense-without-ssr | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 193 B | 193 B | ✓ |
| amp-HASH.js gzip | 308 B | 308 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 2.69 kB | 2.69 kB | ✓ |
| head-HASH.js gzip | 359 B | 359 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.7 kB | 5.7 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.64 kB | 2.64 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 391 B | 391 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 15.9 kB | 15.9 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js dynamic-suspense-without-ssr | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js dynamic-suspense-without-ssr | Change |
| - | - | - | - |
| index.html gzip | 532 B | 532 B | ✓ |
| link.html gzip | 546 B | 546 B | ✓ |
| withRouter.html gzip | 528 B | 528 B | ✓ |
| Overall change | 1.61 kB | 1.61 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js dynamic-suspense-without-ssr | Change |
| - | - | - | - |
| buildDuration | 21.6s | 21.9s | ⚠️ +337ms |
| buildDurationCached | 7.6s | 7.5s | -82ms |
| nodeModulesSize | 475 MB | 475 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | vercel/next.js dynamic-suspense-without-ssr | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 4.912 | 4.81 | -0.1 |
| / avg req/sec | 508.93 | 519.75 | +10.82 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.809 | 1.76 | -0.05 |
| /error-in-render avg req/sec | 1382.03 | 1420.68 | +38.65 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js dynamic-suspense-without-ssr | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 29.2 kB | 29.2 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 73.1 kB | 73.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js dynamic-suspense-without-ssr | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js dynamic-suspense-without-ssr | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 324 B | 324 B | ✓ |
| dynamic-HASH.js gzip | 2.71 kB | 2.71 kB | ✓ |
| head-HASH.js gzip | 357 B | 357 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.79 kB | 5.79 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.76 kB | 2.76 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 392 B | 392 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.1 kB | 16.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js dynamic-suspense-without-ssr | Change |
| - | - | - | - |
| _buildManifest.js gzip | 456 B | 456 B | ✓ |
| Overall change | 456 B | 456 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js dynamic-suspense-without-ssr | Change |
| - | - | - | - |
| index.html gzip | 529 B | 529 B | ✓ |
| link.html gzip | 543 B | 543 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: 0b41fba987728795e3c8bfafe97a339a52478d9c

**Review by kdy1 — DISMISSED** [review]

**Comment by bkyerv** [comment]
what is the fix? what needs to be done for the issue to be fixed? reinstall swc? reinstall latest next?

**Comment by Brooooooklyn** [comment]
@bkyerv you'll need to wait for this pr merged, and the new canary version published. Then install the latest canary version next.js

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 0fdfb7ad15176c342821443dd147c0c0b4343d54

`yarn testheadless test/e2e/dynamic-with-suspense/index.test.ts` 
- dynamic with suspense > should render server-side
- dynamic with suspense > should render client-side

<details>
<summary>Expand output</summary>

  ● dynamic with suspense › should render server-side

    expect(received).toContain(expected) // indexOf

    Expected substring: "Next.js Example"
    Received string:    "<!DOCTYPE html><html><head><style data-next-hide-fouc=\"true\">body{display:none}</style><noscript data-next-hide-fouc=\"true\"><style>body{display:block}</style></noscript><meta charSet=\"utf-8\"/><meta name=\"viewport\" content=\"width=device-width\"/><meta name=\"next-head-count\" content=\"2\"/><noscript data-n-css=\"\"></noscript><script defer=\"\" nomodule=\"\" src=\"/_next/static/chunks/polyfills.js?ts=1652286940821\"></script><script src=\"/_next/static/chunks/webpack.js?ts=1652286940821\" defer=\"\"></script><script src=\"/_next/static/chunks/main.js?ts=1652286940821\" defer=\"\"></script><script src=\"/_next/static/chunks/pages/_app.js?ts=1652286940821\" defer=\"\"></script><script src=\"/_next/static/chunks/pages/_error.js?ts=1652286940821\" defer=\"\"></script><script src=\"/_next/static/development/_buildManifest.js?ts=1652286940821\" defer=\"\"></script><script src=\"/_next/static/development/_ssgManifest.js?ts=1652286940821\" defer=\"\"></script><script src=\"/_next/static/development/_middlewareManifest.js?ts=1652286940821\" defer=\"\"></script><noscript id=\"__next_css__DO_NOT_USE__\"></noscript></head><body><div id=\"__next\" data-reactroot=\"\"></div><script src=\"/_next/static/chunks/react-refresh.js?ts=1652286940821\"></script><script id=\"__NEXT_DATA__\" type=\"application/json\">{\"props\":{\"pageProps\":{\"statusCode\":500}},\"page\":\"/_error\",\"query\":{\"__NEXT_PAGE\":\"/\"},\"buildId\":\"development\",\"isFallback\":false,\"err\":{\"name\":\"Error\",\"message\":\"Invalid suspense option usage in next/dynamic. Read more: https://nextjs.org/docs/messages/invalid-dynamic-suspense\",\"stack\":\"Error: Invalid suspense option usage in next/dynamic. Read more: https://nextjs.org/docs/messages/invalid-dynamic-suspense\\n    at dynamic (webpack-internal:///./node_modules/next/dist/shared/lib/dynamic.js:51:15)\\n    at eval (webpack-internal:///./pages/index.js:14:66)\\n    at Object../pages/index.js (/tmp/next-install-1652286919427/.next/server/pages/index.js:33:1)\\n    at __webpack_require__ (/tmp/next-install-1652286919427/.next/server/webpack-runtime.js:33:42)\\n    at __webpack_exec__ (/tmp/next-install-1652286919427/.next/server/pages/index.js:86:39)\\n    at /tmp/next-install-1652286919427/.next/server/pages/index.js:87:28\\n    at Object.\\u003canonymous\\u003e (/tmp/next-install-1652286919427/.next/server/pages/index.js:90:3)\\n    at Module._compile (node:internal/modules/cjs/loader:1105:14)\\n    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)\\n    at Module.load (node:internal/modules/cjs/loader:981:32)\"},\"gip\":true,\"scriptLoader\":[]}</script></body></html>"

      40 |   it('should render server-side', async () => {
      41 |     const html = await renderViaHTTP(next.url, '/')
    > 42 |     expect(html).toContain('Next.js Example')
         |                  ^
      43 |     expect(html).toContain('Thing')
      44 |   })
      45 |

      at Object.<anonymous> (e2e/dynamic-with-suspense/index.test.ts:42:18)

  ● dynamic with suspense › should render client-side

    expect(received).toBe(expected) // Object.is equality

    Expected: false
    Received: true

      46 |   it('should render client-side', async () => {
      47 |     const browser = await webdriver(next.url, '/')
    > 48 |     expect(await hasRedbox(browser)).toBe(false)
         |                                      ^
      49 |     await browser.close()
      50 |   })
      51 | })

      at Object.<anonymous> (e2e/dynamic-with-suspense/index.test.ts:48:38)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Review by huozhi — COMMENTED** [review]
lgtm on my side

**Review by kdy1 — APPROVED** [review]

**Review by huozhi — APPROVED** [review]

**Pull request merged**